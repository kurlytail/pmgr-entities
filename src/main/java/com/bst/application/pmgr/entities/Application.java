package com.bst.application.pmgr.entities;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import com.bst.configuration.pmgr.entities.PmgrEntitiesConfiguration;

@SpringBootApplication
@ComponentScan(basePackageClasses = { PmgrEntitiesConfiguration.class, Application.class })
@EnableTransactionManagement
public class Application extends SpringBootServletInitializer {

	public static void main(final String[] args) {
		SpringApplication.run(Application.class, args);
	}
}