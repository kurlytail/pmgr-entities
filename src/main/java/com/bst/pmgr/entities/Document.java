package com.bst.pmgr.entities;

import java.util.List;

import javax.persistence.DiscriminatorColumn;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;

import com.bst.pmgr.entities.audit.DocumentAudit;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@EntityListeners(DocumentAudit.class)
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "documentType")
@ToString
@Getter
@EqualsAndHashCode
public class Document {
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;

	@Setter
	private String name;

	@ManyToOne
	@Setter
	private Work work;

	@OneToMany(mappedBy = "document")
	private List<Section> sections;
}
