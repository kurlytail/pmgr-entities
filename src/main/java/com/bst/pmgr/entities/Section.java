package com.bst.pmgr.entities;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.DiscriminatorColumn;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.PrePersist;
import javax.persistence.PreRemove;
import javax.persistence.PreUpdate;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;

import org.springframework.data.annotation.CreatedBy;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedBy;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import com.bst.pmgr.entities.audit.SectionAudit;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnore;

@Entity
@EntityListeners({ SectionAudit.class, AuditingEntityListener.class })
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "sectionType")
public class Section {
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	@JsonIgnore
	private Long id;

	public void setDocument(Document document) {
		this.document = document;
	}

	public void setParentSection(Section parentSection) {
		this.parentSection = parentSection;
	}

	public Long getId() {
		return id;
	}

	public Date getCreatedDate() {
		return createdDate;
	}

	public Date getModifiedDate() {
		return modifiedDate;
	}

	public String getCreatedBy() {
		return createdBy;
	}

	public String getModifiedBy() {
		return modifiedBy;
	}

	public void setChildSections(List<Section> childSections) {
		this.childSections = childSections;
	}

	private String name;
	private String summary;
	private String description;

	public String getSummary() {
		return summary;
	}

	public void setSummary(String summary) {
		this.summary = summary;
	}

	public String getDescription() {
		return description;
	}

	public void setDescription(String description) {
		this.description = description;
	}

	public Document getDocument() {
		return document;
	}

	public Section getParentSection() {
		return parentSection;
	}

	public List<Section> getChildSections() {
		return childSections;
	}

	@ManyToOne
	private Document document;

	@ManyToOne
	private Section parentSection;

	@JsonIgnore
	@OneToMany(mappedBy = "parentSection")
	private List<Section> childSections = new ArrayList<>();

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	@Column(nullable = false, updatable = false)
	@CreatedDate
	@Temporal(TemporalType.TIMESTAMP)
	@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd@HH:mm:ss.SSSZ")
	private Date createdDate;

	@LastModifiedDate
	@Temporal(TemporalType.TIMESTAMP)
	@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd@HH:mm:ss.SSSZ")
	private Date modifiedDate;

    @CreatedBy
    private String createdBy;
 
    @LastModifiedBy
    private String modifiedBy;

    @PrePersist
    @PreUpdate
    @PreRemove
    public void onPrePersist() {
        System.out.println("prepersist");
    }
}

