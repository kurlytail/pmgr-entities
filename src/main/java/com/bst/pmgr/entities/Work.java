package com.bst.pmgr.entities;

import java.util.ArrayList;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.DiscriminatorColumn;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;

import com.bst.pmgr.entities.audit.WorkAudit;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;

@EntityListeners(WorkAudit.class)
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "workType")
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
@Table(name = "WORK_TABLE")
public class Work {	
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;

	@Column(nullable = false)
	private String name;

	@OneToMany(mappedBy = "work")
	private final List<Activity> activities = new ArrayList<>();

	@OneToMany(mappedBy = "work")
	private final List<Document> documents = new ArrayList<>();

	@OneToMany(mappedBy = "work")
	private final List<Tool> tools = new ArrayList<>();

	@OneToMany(mappedBy = "work")
	private final List<Process> processes = new ArrayList<>();

	@OneToMany(mappedBy = "work")
	private final List<ProcessGroup> processGroups = new ArrayList<>();

	@ManyToOne(fetch = FetchType.LAZY)
	private PmgrPerson createdBy;

	@Override
	public boolean equals(final Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (this.getClass() != obj.getClass()) {
			return false;
		}
		final Work other = (Work) obj;
		if (this.activities == null) {
			if (other.activities != null) {
				return false;
			}
		} else if (!this.activities.equals(other.activities)) {
			return false;
		}
		if (this.createdBy == null) {
			if (other.createdBy != null) {
				return false;
			}
		} else if (!this.createdBy.equals(other.createdBy)) {
			return false;
		}
		if (this.documents == null) {
			if (other.documents != null) {
				return false;
			}
		} else if (!this.documents.equals(other.documents)) {
			return false;
		}
		if (this.id == null) {
			if (other.id != null) {
				return false;
			}
		} else if (!this.id.equals(other.id)) {
			return false;
		}
		if (this.name == null) {
			if (other.name != null) {
				return false;
			}
		} else if (!this.name.equals(other.name)) {
			return false;
		}
		if (this.processGroups == null) {
			if (other.processGroups != null) {
				return false;
			}
		} else if (!this.processGroups.equals(other.processGroups)) {
			return false;
		}
		if (this.processes == null) {
			if (other.processes != null) {
				return false;
			}
		} else if (!this.processes.equals(other.processes)) {
			return false;
		}
		if (this.tools == null) {
			if (other.tools != null) {
				return false;
			}
		} else if (!this.tools.equals(other.tools)) {
			return false;
		}
		return true;
	}

	public List<Activity> getActivities() {
		return this.activities;
	}

	public PmgrPerson getCreatedBy() {
		return this.createdBy;
	}

	public List<Document> getDocuments() {
		return this.documents;
	}

	public Long getId() {
		return this.id;
	}

	public String getName() {
		return this.name;
	}

	public List<Process> getProcesses() {
		return this.processes;
	}

	public List<ProcessGroup> getProcessGroups() {
		return this.processGroups;
	}

	public List<Tool> getTools() {
		return this.tools;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((this.activities == null) ? 0 : this.activities.hashCode());
		result = prime * result + ((this.createdBy == null) ? 0 : this.createdBy.hashCode());
		result = prime * result + ((this.documents == null) ? 0 : this.documents.hashCode());
		result = prime * result + ((this.id == null) ? 0 : this.id.hashCode());
		result = prime * result + ((this.name == null) ? 0 : this.name.hashCode());
		result = prime * result + ((this.processGroups == null) ? 0 : this.processGroups.hashCode());
		result = prime * result + ((this.processes == null) ? 0 : this.processes.hashCode());
		result = prime * result + ((this.tools == null) ? 0 : this.tools.hashCode());
		return result;
	}

	public void setCreatedBy(final PmgrPerson createdBy) {
		this.createdBy = createdBy;
	}

	public void setName(final String name) {
		this.name = name;
	}

	@Override
	public String toString() {
		return "Work [id=" + this.id + ", name=" + this.name + ", createdBy=" + this.createdBy + "]";
	}

}
