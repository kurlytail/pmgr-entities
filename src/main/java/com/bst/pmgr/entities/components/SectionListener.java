package com.bst.pmgr.entities.components;

import javax.persistence.PrePersist;

import org.springframework.stereotype.Component;

import com.bst.pmgr.entities.Section;

@Component
public class SectionListener {
	
	@PrePersist
	public void prePersist(final Section section) throws Exception {
		if (section.getDocument() == null) {
			throw new IllegalStateException("Section.document should be set");
		}
		
		if (section.getParentSection() != null) {
			if (section.getDocument() != section.getParentSection().getDocument()) {
				throw new IllegalStateException("Section.parentSection.document must be the same as Section.document");
			}
		}
	}
}
