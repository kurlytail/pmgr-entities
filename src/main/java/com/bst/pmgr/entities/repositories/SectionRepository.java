package com.bst.pmgr.entities.repositories;

import java.util.List;

import org.springframework.data.repository.CrudRepository;
import org.springframework.stereotype.Repository;

import com.bst.pmgr.entities.Section;

@Repository
public interface SectionRepository<T extends Section> extends CrudRepository<T, Long> {
	  public List<T> findAll();
}
