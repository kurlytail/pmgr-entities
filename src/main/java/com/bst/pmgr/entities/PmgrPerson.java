package com.bst.pmgr.entities;

import java.util.ArrayList;
import java.util.List;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.Id;
import javax.persistence.OneToMany;
import javax.validation.constraints.NotNull;

import com.bst.user.authentication.entities.Person;
import com.bst.utility.components.AuditListener;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;

@EntityListeners(AuditListener.class)
@Entity
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
public class PmgrPerson {
	@Id
	private Long id;

	@NotNull
	@Column(updatable = false)
	private Person person;

	@OneToMany(cascade = CascadeType.ALL, orphanRemoval = true, mappedBy = "createdBy")
	private final List<Work> works = new ArrayList<>();

	public void addWork(final Work work) {
		this.works.add(work);
		work.setCreatedBy(this);
	}

	@Override
	public boolean equals(final Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (this.getClass() != obj.getClass()) {
			return false;
		}
		final PmgrPerson other = (PmgrPerson) obj;
		if (this.id == null) {
			if (other.id != null) {
				return false;
			}
		} else if (!this.id.equals(other.id)) {
			return false;
		}
		if (this.person == null) {
			if (other.person != null) {
				return false;
			}
		} else if (!this.person.equals(other.person)) {
			return false;
		}
		if (this.works == null) {
			if (other.works != null) {
				return false;
			}
		} else if (!this.works.equals(other.works)) {
			return false;
		}
		return true;
	}

	public Long getId() {
		return this.id;
	}

	public Person getPerson() {
		return this.person;
	}

	public List<Work> getWorks() {
		return this.works;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((this.id == null) ? 0 : this.id.hashCode());
		result = prime * result + ((this.person == null) ? 0 : this.person.hashCode());
		result = prime * result + ((this.works == null) ? 0 : this.works.hashCode());
		return result;
	}

	public void removeWork(final Work work) {
		this.works.remove(work);
	}

	public void setId(final Long id) {
		this.id = id;
	}

	public void setPerson(final Person person) {
		this.person = person;
	}

	@Override
	public String toString() {
		return "PmgrPerson [id=" + this.id + ", person=" + this.person + "]";
	}

}
