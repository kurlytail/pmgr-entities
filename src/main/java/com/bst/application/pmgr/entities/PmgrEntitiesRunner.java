package com.bst.application.pmgr.entities;

import java.util.Random;

import javax.persistence.EntityManager;
import javax.transaction.Transactional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.CommandLineRunner;
import org.springframework.stereotype.Service;

import com.bst.pmgr.entities.Work;
import com.bst.pmgr.entities.repositories.WorkRepository;

@Service
@Transactional
public class PmgrEntitiesRunner implements CommandLineRunner {

	@Autowired
	private WorkRepository workRepository;

	@Autowired
	private EntityManager entityManager;

	@Override
	public void run(final String... args) throws Exception {

		final long countOfWork = Math.abs(new Random().nextLong()) % 10;
		long indexOfWork = 0;
		
		while (indexOfWork < countOfWork) {
			final Work work = new Work();
			this.workRepository.save(work);
			++indexOfWork;
		}

		this.entityManager.flush();
	}

}
