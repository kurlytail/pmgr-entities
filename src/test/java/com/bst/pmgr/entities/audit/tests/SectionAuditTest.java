package com.bst.pmgr.entities.audit.tests;

import java.util.List;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.domain.EntityScan;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.EnableAspectJAutoProxy;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.TestExecutionListeners;
import org.springframework.test.context.TestExecutionListeners.MergeMode;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import com.bst.pmgr.entities.Document;
import com.bst.pmgr.entities.PmgrPerson;
import com.bst.pmgr.entities.Section;
import com.bst.pmgr.entities.Work;
import com.bst.pmgr.entities.components.SectionListener;
import com.bst.pmgr.entities.repositories.SectionRepository;
import com.bst.user.authentication.entities.Person;
import com.bst.utility.components.RepositoryAspect;
import com.bst.utility.testlib.SnapshotListener;

@ExtendWith(SpringExtension.class)
@DataJpaTest
@TestExecutionListeners(listeners = SnapshotListener.class, mergeMode = MergeMode.MERGE_WITH_DEFAULTS)
@ContextConfiguration(classes = { TestConfiguration.class })
@TestPropertySource("classpath:section-audit-test.properties")
public class SectionAuditTest {

	@TestConfiguration
	@EntityScan("com.bst.pmgr.entities")
	@EnableJpaRepositories("com.bst.pmgr.entities.repositories")
	@EnableAspectJAutoProxy
	static class SectionAuditTestConfiguration {
		@Bean
		public RepositoryAspect getRepositoryAspect() {
			return new RepositoryAspect();
		}

		@Bean
		public SectionListener getSectionListener() {
			return new SectionListener();
		}
	}

	@Autowired
	private TestEntityManager entityManager;

	@Autowired
	private SectionRepository<Section> sectionRepository;

	@SuppressWarnings("unused")
	@Autowired
	private SectionListener sectionListener;

	@Test
	public void testDocumentField() throws Exception {

		final Work work = new Work();
		work.setName("some name");

		final PmgrPerson pmgrPerson = new PmgrPerson();
		final Person person = new Person();
		pmgrPerson.setPerson(person);
		pmgrPerson.setName("test person");
		work.setPerson(pmgrPerson);

		final Section section = new Section();
		section.setName("testName");
		Assertions.assertThrows(RuntimeException.class, () -> {
			this.sectionRepository.save(section);
			this.entityManager.flush();
		});

		List<Section> sections = this.sectionRepository.findAll();
		SnapshotListener.expect(sections).toMatchSnapshot();

		final Document document = new Document();
		document.setName("Document");
		document.setMetaName("projectStaffAssignments");
		work.addToDocuments(document);
		document.addToTopLevelSections(section);

		this.sectionRepository.save(section);
		this.entityManager.flush();

		sections = this.sectionRepository.findAll();
		SnapshotListener.expect(sections).toMatchSnapshot();
	}

	@Test
	public void testParentSectionField() throws Exception {

		final Work work = new Work();
		work.setName("some test");

		final PmgrPerson pmgrPerson = new PmgrPerson();
		final Person person = new Person();
		pmgrPerson.setPerson(person);
		pmgrPerson.setName("test person");
		work.setPerson(pmgrPerson);

		final Document document = new Document();
		document.setName("document");
		document.setMetaName("projectStaffAssignments");
		work.addToDocuments(document);

		final Document document1 = new Document();
		document1.setName("document1");
		document1.setMetaName("projectStaffAssignments");
		work.addToDocuments(document1);

		final Section section = new Section();
		section.setName("testName");
		document.addToTopLevelSections(section);

		Section parentSection = new Section();
		parentSection.setName("parent testName");
		document1.addToTopLevelSections(parentSection);

		parentSection = this.sectionRepository.save(parentSection);
		parentSection.addToChildSections(section);

		document.addToTopLevelSections(section);
		this.sectionRepository.save(section);

		List<Section> sections = this.sectionRepository.findAll();
		SnapshotListener.expect(sections).toMatchSnapshot();

		parentSection.setDocument(document1);
		parentSection = this.sectionRepository.save(parentSection);
		section.setParentSection(parentSection);
		this.sectionRepository.save(section);
		this.entityManager.flush();

		sections = this.sectionRepository.findAll();
		SnapshotListener.expect(sections).toMatchSnapshot();

	}

}
