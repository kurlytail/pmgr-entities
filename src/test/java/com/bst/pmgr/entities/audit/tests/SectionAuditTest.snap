{
  "testDocumentField" : [ [ ], [ {
    "name" : "testName",
    "document" : {
      "name" : "Document"
    }
  } ] ],
  "testParentSectionField" : [ [ {
    "name" : "parent testName",
    "document" : {
      "name" : "document1"
    }
  } ], [ {
    "name" : "parent testName",
    "document" : {
      "name" : "document"
    }
  }, {
    "name" : "testName",
    "document" : {
      "name" : "document"
    },
    "parentSection" : {
      "name" : "parent testName",
      "document" : {
        "name" : "document"
      }
    }
  } ] ]
}